File: 503.next-greater-element-ii.java
public class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int[] r = new int[nums.length];
        if (nums.length == 0)
            return r;
        Arrays.fill(r, -1);
        int[] N = new int[nums.length];
        int[] P = new int[nums.length];
        N[0] = nums[0];
        P[0] = 0;
        int p = 0, l = 2 * r.length;
        for (int i = 1; i < l - 1; ++i) {
            int ri = i % r.length;
            if (nums[ri] <= N[p]) {
                if (++p >= r.length)
                    break;
                N[p] = nums[ri];
                P[p] = ri;
            } else {
                while (p >= 0 && N[p] < nums[ri]) {
                    r[P[p]] = nums[ri];
                    --p;
                }
                N[++p] = nums[ri];
                P[p] = ri;
            }
        }
        return r;
    }
}

File: 503.next-greater-element-ii.js
/*
 * @lc app=leetcode id=503 lang=javascript
 *
 * [503] Next Greater Element II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[]}
 */

// Approach #3 Using Stack
var nextGreaterElements = function (nums) {
  let res = new Array(nums.length).fill(0);
  let stack = [];
  // 为什么是2 * nums.length
  for (let i = 2 * nums.length - 1; i >= 0; --i) {
    console.log(stack);
    while (stack.length !== 0 && nums[stack[0]] <= nums[i % nums.length]) {
      stack.shift();
    }
    res[i % nums.length] = stack.length === 0 ? -1 : nums[stack[0]];
    console.log(res);
    stack.unshift(i % nums.length);
  }

  return res;
};

nextGreaterElements([4, 1, 2, 3]);

// [1,2,1,1,2,1]
// @lc code=end
